doc/arch/006-forward-compatibility-for-enums.md

```diff
@@ -0,0 +1,55 @@
+# ADR: Forward compatibility for persistable enums from Server responses
+
+
+## Status
+
+Accepted
+
+
+## Context
+
+RESTful APIs tend to return enumerated values to represent properties with a fixed number of well-defined values. These are usually represented as `enum` classes on the client. When new business use cases are introduced, there could be additions to these values. If these values are consumed on the fly and are not persisted on the client, they could simply be treated as a special case (unknown) or ignored all together.
+
+In Simple, some of these values have to be persisted in the database. Hence, these enumerated values cannot be represented using a Kotlin `enum` class.
+
+
+## Decision
+
+To maintain forward-compatibility with enum values, we have to roll out our own infrastructure. To accommodate, we use sealed classes to represent enumerated values so that,
+
+- We could save these values to the database.
+- Use these values in newer versions of the app.
+- Send the same enum values back to the server during sync even though the current version of the app cannot interpret these values during runtime.
+
+In order or achieve this, we require 4 classes.
+
+1. A sealed class hierarchy that is used to represent the enumerated values. Which also includes an `Unknown(val actualValue: String)` data class which is used to capture and handle newly introduced values.
+2. An implementation of the `SafeEnumTypeAdapter` class.
+3. A `RoomTypeConverter` to safely store and retrieve newly added enum values.
+4. A `MoshiTypeAdapter` to serialize and deserialize newly added enum values.
+
+The interesting part is the `SafeEnumTypAdapter` class definition,
+
+```kotlin
+open class SafeEnumTypeAdapter<T>(
+    val knownMappings: Map<T, String>,
+    private val unknownEnumToStringConverter: (T) -> String,
+    private val unknownStringToEnumConverter: (String) -> T
+) {
+  // …
+}
+```
+
+The `knownMappings` field takes in a `Map` that can convert to and from enumerations that the current version of the application understands.
+
+The `unknownEnumToStringConverter` converts strings to their appropriate sealed class representations.
```

Comment:
Isn't this the other way around? This is to convert unknown enum values to their actual string representation (usually used when pushing data to the server).